### PR TITLE
feat(react): add `componentDidCatch`

### LINKED ISSUE #222: feat: Error boundaries
https://github.com/aidenybai/million-react/issues/6

### PR DESCRIPTION
**Pull request by QuiiBz**
**Please describe the changes this PR makes and why it should be merged:**

Closes #222
Relates to https://github.com/aidenybai/million-react/issues/6

In draft.

Currently, `componentDidCatch` in class components only works for render errors, not for effects. It's a lot harder to achieve considering the current architecture of Million, because we need to retrieve recursively the parent VNode of a VNode, until is a class component and has `componentDidCatch`.

**UPDATE**
I found a way to catch all errors from effects, but it might add a (lot?) of overhead. In each VNode, we have to keep track of the parent VNode (`_parent`). Class Components also add a `_component` field to the root VNode. When an error is thrown from an effect (initial or subsequent), we traverse up the three of VNodes until we find one that has a `_component`, and run its `componentDidCatch`. This is similar to Preact's implementation.

**Status**

- [x] Code changes have been tested against prettier, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**

- [x] This PR changes the codebase
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
  - [x] This PR changes the internal workings with no modifications to the external API (bug fixes, performance improvements)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by aidenybai** [comment]
This looks like the best solution for classes. LGTM, merge when you're ready!

**Comment by QuiiBz** [comment]
I'll research a bit more about catching errors from effects before merging, because `componentDidCatch` currently only works for (re)render errors.

**Comment by QuiiBz** [comment]
Found an issue with the changes inside the Vite plugin, it was throwing an error and such fail to transform the AST.

Fixed by only adding the `_parent`  to the root velement of a class/function component instead of every velement (previously done in `m`), and remove the faulty code inside vite-plugin-million.

Left: this PR
Right: main branch

Screenshot 2022-08-06 at 08 58 12

File: packages/million/render.ts [hunk 1]
```diff
@@ -22,6 +22,9 @@ export const patch = (
   hook: Hook = () => true,
   effects: Effect[] = [],
 ): DOMNode => {
+  if (typeof newVNode === 'object' && typeof oldVNode === 'object') {
+    newVNode._parent = oldVNode._parent;
```
Comment (aidenybai) [inline packages/million/render.ts:26]: opt in logic?

**Comment by aidenybai** [comment]
Congrats on the ship!!

**Pull request merged**